README.rst
Installation instructions, until this is part of Jmbo:

Add jmbo-janrain to your buildout as part of the buildout eggs::

    [buildout]

    extensions=
        ...
        mr.developer

    eggs=
        ...
        jmbo-janrain

    [sources]
    jmbo-janrain = git git://github.com/praekelt/jmbo-janrain.git branch=develop

Add it to your INSTALLED_APPS::

    INSTALLED_APPS = (
    ...
    'janrain',
    ...
    )

Add the middleware::

    middleware_classes = (
    ...
    'janrain.middleware.JanrainMiddleware',
    ...
    )

Add the following parameters to your settings file::

    # URL given as your domain url by Janrain, including the https parts.
    JANRAIN_URL = 'JANRAIN_URL'
    
    # The janrain client ID, shown on the Janrain site.
    JANRAIN_CLIENT_ID = 'JANRAIN_CLIENT_ID'

    # The Janrain client secret, shown on the Janrain site.
    JANRAIN_CLIENT_SECRET = 'JANRAIN_CLIENT_SECRET'

The product includes a South migration, so I found that the following works when using the standard Praekelt setup:

* Backup.

* Run a level 2 deploy. This will pull in the product and dependencies. The server will either not start up or will throw 503 errors, since the migration is not run yet.

* Do a South migration::
    ./bin/{yoursite}-qa-web-site migrate janrain

* Run a level 1 install. Everything should now start up. You might want to add the stuff above into all the settings files.

Find the complete set of `Jmbo docs here <http://jmbo.readthedocs.org/>`_



CHANGELOG.rst
Changelog
=========

0.1
---
#. Reduce calls to `reverse`.
#. Guard against no user being present on the request.

0.0.1
-----
#. Initial product



janrain/tests.py
import unittest

from django.contrib.sites.models import Site
from django.core.urlresolvers import reverse
from django.http import Http404
from django.test import TestCase
from django.test.client import Client


janrain/middleware.py
import re

from django.core.urlresolvers import reverse
from django.conf import settings

from preferences import preferences

from foundry.models import Member
from janrain.models import JanrainProfile
from janrain.models import STATUS_DIRTY, STATUS_SYNCED


# Pre-compute URL reversals since they can't change during Django process
PATHS = (reverse('complete-profile'), reverse('edit-profile'))


class JanrainMiddleware:

    def process_request(self, request):
        if request.META['PATH_INFO'] in PATHS:
            user = getattr(request, 'user', None)
            if (user is not None) and user.is_authenticated():
                # Possible error in JanrainProfile model. User is a foreign key
                # and not a one-to-one.
                janrain_profile = user.janrainprofile_set.all()[0]
                if request.META['REQUEST_METHOD'] == 'POST':
                    janrain_profile.status = STATUS_DIRTY
                    print 'status set to dirty'
                    janrain_profile.save()
                if request.META['REQUEST_METHOD'] == 'GET':
                    janrain_profile.status = STATUS_DIRTY
                    print 'Will sync now'
                    # TODO: update profile from Janrain data.
                    janrain_profile.status = STATUS_SYNCED
                    janrain_profile.save()


janrain/retrieve_data.py
"""
Reference and playing around with the janrain api
"""
import requests
import sys
import pprint

# Get imports and errors right for json/simplejson
try:
    import simplejson as json
except ImportError:
    import json

try:
    JSONDecodeError = json.JSONDecodeError
except AttributeError:
    JSONDecodeError = ValueError


client_id = "qvufhhghk4ccgrbqcq8cap2rm6p6kgnz"
client_secret = "v5b9srfzsnd7wnrpq2zv4mdyete6js8q"
# janrain_url = "login.kagisodigital.com"
janrain_url = "https://kagiso.dev.janraincapture.com"
api_call = "entity.count"
type_name = "user"

payload = {'client_id': client_id,
           'client_secret': client_secret,
           'type_name': type_name,
           }
response = requests.post("%s/%s" % (janrain_url, api_call), data=payload)
print response
try:
    pprint.pprint(response.json())
except:
    pprint.pprint(response.content)
api_call = 'entity.create'
user_attributes = {'givenName':'Bob', 'familyName': 'Smith'}
print json.dumps(user_attributes)
payload ['attributes'] = json.dumps(user_attributes)
response = requests.post("%s/%s" % (janrain_url, api_call), data=payload)
print response.content
import pdb;pdb.set_trace()
sys.exit()


janrain/migrations/0001_initial.py
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'JanrainProfile'
        db.create_table('janrain_janrainprofile', (
            ('id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('user', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['auth.User'], unique=True)),
            ('janrain_uuid', self.gf('django.db.models.fields.CharField')(max_length=128, null=True, blank=True)),
        ))
        db.send_create_signal('janrain', ['JanrainProfile'])


    def backwards(self, orm):
        # Deleting model 'JanrainProfile'
        db.delete_table('janrain_janrainprofile')


    models = {
        'auth.group': {
            'Meta': {'object_name': 'Group'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        'auth.permission': {
            'Meta': {'ordering': "('content_type__app_label', 'content_type__model', 'codename')", 'unique_together': "(('content_type', 'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Group']", 'symmetrical': 'False', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        'janrain.janrainprofile': {
            'Meta': {'object_name': 'JanrainProfile'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'janrain_uuid': ('django.db.models.fields.CharField', [], {'max_length': '128', 'null': 'True', 'blank': 'True'}),
            'user': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']", 'unique': 'True'})
        }
    }

    complete_apps = ['janrain']

janrain/migrations/0002_auto__add_field_janrainprofile_last_synced__add_field_janrainprofile_s.py
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding field 'JanrainProfile.last_synced'
        db.add_column('janrain_janrainprofile', 'last_synced',
                      self.gf('django.db.models.fields.DateTimeField')(auto_now_add=True, null=True, blank=True),
                      keep_default=False)

        # Adding field 'JanrainProfile.status'
        db.add_column('janrain_janrainprofile', 'status',
                      self.gf('django.db.models.fields.PositiveIntegerField')(default=0),
                      keep_default=False)


    def backwards(self, orm):
        # Deleting field 'JanrainProfile.last_synced'
        db.delete_column('janrain_janrainprofile', 'last_synced')

        # Deleting field 'JanrainProfile.status'
        db.delete_column('janrain_janrainprofile', 'status')


    models = {
        'auth.group': {
            'Meta': {'object_name': 'Group'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        'auth.permission': {
            'Meta': {'ordering': "('content_type__app_label', 'content_type__model', 'codename')", 'unique_together': "(('content_type', 'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Group']", 'symmetrical': 'False', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        'janrain.janrainprofile': {
            'Meta': {'object_name': 'JanrainProfile'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'janrain_uuid': ('django.db.models.fields.CharField', [], {'max_length': '128', 'null': 'True', 'blank': 'True'}),
            'last_synced': ('django.db.models.fields.DateTimeField', [], {'auto_now_add': 'True', 'null': 'True', 'blank': 'True'}),
            'status': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'user': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']", 'unique': 'True'})
        }
    }

    complete_apps = ['janrain']

janrain/models.py
from datetime import datetime
import requests

from django.db import models
from django.db.models.signals import post_save

from django.contrib.auth import user_logged_in
from django.contrib.auth.models import User
from django.dispatch import receiver
from django.conf import settings
from foundry.models import Member

# Get imports and errors right for json/simplejson
try:
    import simplejson as json
except ImportError:
    import json

try:
    JSONDecodeError = json.JSONDecodeError
except AttributeError:
    JSONDecodeError = ValueError

# Janrain client id and secret
JANRAIN_URL = settings.JANRAIN_URL

STATUS_SYNCED = 1
STATUS_DIRTY = 2
STATUS_CONFLICT = 4

base_payload = {'client_id': settings.JANRAIN_CLIENT_ID,
           'client_secret': settings.JANRAIN_CLIENT_SECRET,
           'type_name': 'user',
           }

# Mapping user and profile fields to Janrain equivalents. Each tuple is in the
# form (jmbo_name, janrain_name).
field_mappings = (
    ('first_name', 'givenName'),
    ('last_name', 'familyName'),
    ('email', 'email'),
    ('username', 'displayName'),
    )

# TODO: Provide a way for the field mappings to be overridden from settings.py

def map_user_to_janrain(user):
    """
    Given a user object, provide a dictionary of mapped fields and values to
    give to Janrain to update the user.
    """
    maps = {x[0]:x[1] for x in field_mappings}
    attributes = {}
    user_fields = ('first_name', 'last_name', 'username', 'email')

    for field in user_fields:
        value = getattr(user, field, None)
        if value:
            attributes[maps[field]] = value

    return attributes

#TODO: This is for reference only. Remove.
"""
    first_name = models.CharField(_('first name'), max_length=30, blank=True)
    username = models.CharField(_('username'), max_length=30, unique=True,
        help_text=_('Required. 30 characters or fewer. Letters, numbers and '
                    '@/./+/-/_ characters'))
    last_name = models.CharField(_('last name'), max_length=30, blank=True)
    email = models.EmailField(_('e-mail address'), blank=True)
    last_login = models.DateTimeField(_('last login'), default=timezone.now)
    date_joined = models.DateTimeField(_('date joined'), default=timezone.now)

from prefs:
last_login
date_joined
image
date_taken
view_count
crop_from
effect
facebook_id
twitter_username
address
city
zipcode
province
dob
gender
about_me
mobile_number
receive_sms
receive_email
country
is_profile_complete
    """

@receiver(user_logged_in)
def on_user_logged_in(sender, **kwargs):
    """
    Receiver for when the user logs in. It's placed here on recommendation of
    the Django docs.
    """
    print 'user logged in'
    user = kwargs['user']

    # Get or create the user's Janrain profile
    janrain_profile, created = JanrainProfile.objects.get_or_create(user=user)

    if not janrain_profile:
        janrain_profile = JanrainProfile(user=user)

    # TODO: The following parts may be slow, since the calls are synchronous.
    # Consider queueing.

    # Try to find a user on Janrain to tie to our own user.
    # If no corresponding Janrain user exists, create one.
    # TODO: For now, we don't try to tie these up. We just create a user if we
    # don't have a uuid.
    payload = base_payload.copy()
    if janrain_profile.janrain_uuid:
        # Just update the last logged in time on Janrain. TODO
        pass
    else:
        # Create a Janrain user.
        user_attributes = map_user_to_janrain(user)
        payload ['attributes'] = json.dumps(user_attributes)
        response = requests.post("%s/entity.create" % JANRAIN_URL, data=payload)
        # TODO: Handle response codes other than 200
        struct = json.loads(response.content)
        if struct['stat'] == 'ok':
            janrain_profile.janrain_uuid = struct['uuid']

    janrain_profile.save()    
    # Populate the profile of the logged-in user, if possible.

@receiver(post_save, sender=Member)
def on_user_profile_saved(sender, **kwargs):
    """
    Receiver for when the user profile is saved. Push to Janrain.
    """

    # On initial profile save during registration, just return.
    janrain_profiles = JanrainProfile.objects.filter(user=kwargs['instance'])
    if len(janrain_profiles) == 0:
        print 'Initial profile save'
        return

    janrain_profile = janrain_profiles[0]
    
    # The dirty status is set on profile form save. This prevents spurious
    # overwrites and allows us to retrigger an update if this fails.
    if janrain_profile.status == STATUS_DIRTY:
        payload = base_payload.copy()
        payload['uuid'] = janrain_profile.janrain_uuid
        user_attributes = map_user_to_janrain(janrain_profile.user)
        payload['value'] = json.dumps(user_attributes)
        response = requests.post("%s/entity.update" % JANRAIN_URL, data=payload)
        # TODO: Handle response codes other than 200
        struct = json.loads(response.content)
        if struct['stat'] == 'ok':
            janrain_profile.status = STATUS_SYNCED
            janrain_profile.last_synced = datetime.now()
            #TODO: last synced time set
            janrain_profile.save()


class JanrainProfile(models.Model):

    user = models.ForeignKey(
        User, 
        unique=True,
    )
    
    janrain_uuid = models.CharField(
        max_length=128,
        blank=True, 
        null=True,
    )

    last_synced = models.DateTimeField(
        auto_now_add=True,
        null=True,
    )

    # Status indicator. Might be refactored to flags.
    status = models.PositiveIntegerField(
        default=0, 
    )

    def __unicode__(self):
        return self.user.username
    
